docs/detailed-design-doc.rst

```diff
@@ -0,0 +1,357 @@
+  of the workloads.
+- Multiple workloads can be mixed to better emulate production traffic where
+  multiple groups of query patterns exist.
+- Accurate and precise control of the query (a.k.a. event) rates for each
+  workload.
+- Live monitoring of throughput and latency statistics.
+- Logging of throughput and latency time series during the benchmark run in
+  standardized data formats.
+- Standardized plots to visualize the results post-benchmarking.
+
+mybench design
+==============
+
+Internal architecture
+---------------------
+
+mybench is a Golang library that enables developers to create, run, and analyze
+benchmarks. To minimize the efforts required to define a benchmark, the
+top-level API is kept small and much of the complexities of running a benchmark
+is implemented within mybench. Figure 1 depicts the internal architecture of
+mybench.
+
+.. figure:: images/architecture.svg
+
+   Figure 1: mybench architecture showing 3 ``Workload``\s each running with two
+   ``BenchmarkWorker``\s
+
+mybench defines a **Benchmark** to consists of multiple **Workloads**. At least
+one Workload must be defined for a Benchmark. Each Workload is defined by its
+``Event`` callback, within which contains the code being benchmarked. For
+example, the ``Event`` callback for one workload could implement a sequence of
+queries that is typical for a web request to a particular API endpoint while
+the ``Event`` callback for another workload could implement a different
+sequence of queries that is typical for a different API endpoint. In the
+figure, three such workloads are mixed together during the benchmark. Each
+workload is configured by the number of parallel ``BenchmarkWorker``\s executing
+the events and an overall event rate. In the figure, all three workloads are
+configured to be executed with two parallel ``BenchmarkWorker``\s. The splitting
+of workload into multiple parallel workers is useful for simulating scenarios
+where many connections are sending queries to the database simultaneously.
+``BenchmarkWorker``\s are implemented as goroutines each with its own connection
+to the database. The overall event rate is equally divided between all
+``BenchmarkWorker``\s. Each ``BenchmarkWorker`` embeds a ``Looper`` object,
+which precisely controls the event rate (see `Rate control via
+"temporal-discretization" looper`_), and an ``OnlineHistogram`` object, which
+stores the throughput and latency statistics during the benchmark run. Data
+held inside the ``OnlineHistogram`` objects are periodically read by the
+``DataLogger`` (see `High-performance data logging via double buffering`_).
+Once the data is read, the ``DataLogger`` aggregates the throughput and latency
+statistics across all ``BenchmarkWorker``\s and all ``Workload``\s into
+per-workload and overall throughput and latency statistics. These values are
+then written into a SQLite database.
+
+Users developing a custom benchmark do not need to be concerned with the
```

Comment:
Consider splitting this section into a usage section rather then leaving it in the "internals" section.

Reply:
I'm going to leave this here as I don't have an usage section.